yesod-test/test/main.hs

```diff
@@ -575,9 +578,44 @@ main = hspec $ do
             statusIs 200
             (requireJSONResponse :: YesodExample site [Text]) `liftedShouldThrow` (\(e :: SomeException) -> True)
 
+    describe "Checks the order of addPostParam arguments addition" $ yesodSpec defaultRoutedApp $ do
+        yit "checks whether the form fails with bad input" $ do
+            get TupleR
+
+            request $ do
+              setUrl TupleR
+              addToken
+              setMethod "POST"
+              addPostParam "data" "ABC"
+              addPostParam "data" "ABC"
+            statusIs 400
+        yit "checks whether the order of addPostParam addition is the same as in the browser" $ do  -- See: https://github.com/yesodweb/yesod/issues/1846
+            get TupleR
+           
+            request $ do
+              setUrl TupleR
+              addToken
+              setMethod "POST"
+              addPostParam "data" "100"
+              addPostParam "data" "ABC"
+            statusIs 200
+        yit "checks whether the order of addPostParam addition is not backwards" $ do  -- See: https://github.com/yesodweb/yesod/issues/1846
+            get TupleR
+
+            request $ do
+              addToken
+              setUrl TupleR
+              setMethod "POST"
+              addPostParam "data" "ABC"
+              addPostParam "data" "100"
+            statusIs 400
```

Comment:
With nicer formatting, and by squashing the two failure cases, I think we could reduce this to something like:

```suggestion
    -- See: https://github.com/yesodweb/yesod/issues/1846
    describe "addPostParam" $ yesodSpec defaultRoutedApp $ do

        yit "ensures parameters are applied in the right order" $ do
            get TupleR

            request $ do
              addToken
              setUrl TupleR
              setMethod "POST"
              addPostParam "data" "ABC"
              addPostParam "data" "100"

            statusIs 400

            get TupleR

            request $ do
              setUrl TupleR
              addToken
              setMethod "POST"
              addPostParam "data" "100"
              addPostParam "data" "ABC"

            statusIs 200
```